### PR TITLE
PeerMessageBus refactor

### PR DESCRIPTION
**Pull request by AlanVerbner**
## Description

Peer message bus implementation changed in order to:

- Don't send the notification twice if a subscription is made with two `MessageClassifier` with same message code, eg: `MessageClassifier(1,2,3)` and `MessageClassifier(2)` should only trigger once
- Allow users to Unsubscribe by code message code instead of the entire `MessageClassifier`

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala [hunk 1]
```diff
@@ -27,57 +27,64 @@ object PeerMessageBusActor {
...
-  case class Subscription(subscriber: ActorRef, classifier: MessageClassifier)
-
   class PeerMessageBus extends ActorEventBus {
 
     override type Event = MessageFromPeer
     override type Classifier = MessageClassifier
 
-    private var subscriptions: Seq[Subscription] = Nil
+    //FIXME Remove this var
+    private var subscriptions: Map[(Subscriber, PeerSelector), Set[Int]] = Map.empty
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: should we remove this var now?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: We can but I rather to do it on another PR as it was not introduced in this fix. 

@adamsmo do you think is worthy to do it now? :innocent:

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: ok, I think it is worth creating task to not forgot about this

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: done `EC-197`

**Comment by adamsmo** [comment]
LGTM

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala [hunk 2]
```diff
@@ -27,57 +27,64 @@ object PeerMessageBusActor {
...
-      if (subscriptions.contains(subscription)) {
-        subscriptions = subscriptions.filterNot(_ == subscription)
-        true
-      } else {
-        false
+      subscriptions.get((subscriber, from.peerSelector)).exists { messageCodes =>
+        val newMessageCodes = messageCodes -- from.messageCodes
+        if (messageCodes == newMessageCodes) false
+        else {
+          if(newMessageCodes.isEmpty) subscriptions = subscriptions - (subscriber -> from.peerSelector)
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: I think changing `from.peerSelector` to `messageCodes` would make this easier to read

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: maybe `(subscriber, from.peerSelector)` to be consistent?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: Fixed!

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala [hunk 3]
```diff
@@ -27,57 +27,64 @@ object PeerMessageBusActor {
...
-      if (subscriptions.contains(subscription)) {
-        subscriptions = subscriptions.filterNot(_ == subscription)
-        true
-      } else {
-        false
+      subscriptions.get((subscriber, from.peerSelector)).exists { messageCodes =>
+        val newMessageCodes = messageCodes -- from.messageCodes
+        if (messageCodes == newMessageCodes) false
+        else {
+          if(newMessageCodes.isEmpty) subscriptions = subscriptions - ((subscriber, from.peerSelector))
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: `((subscriber, from.peerSelector))` <- are double parentheses required?

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala (hunk 1 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: I think this var doesn't hurt much as long as it's encapsulated in an actor

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala [hunk 4]
```diff
@@ -27,57 +27,64 @@ object PeerMessageBusActor {
...
 
     override def subscribe(subscriber: ActorRef, to: Classifier): Boolean = {
-      val subscription = Subscription(subscriber, to)
-      if (subscriptions.contains(subscription)) {
-        false
-      } else {
-        subscriptions = subscriptions :+ subscription
+      val newSubscriptions = subscriptions.get((subscriber, to.peerSelector)) match {
+        case Some(messageCodes) =>
+          if (messageCodes == to.messageCodes) subscriptions
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: this `if` isn't really needed, since it's a `Set` we can just return `subscriptions + ((subscriber, to.peerSelector) -> (messageCodes ++ to.messageCodes))`, can't we?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: you are right! fixed.

File: src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala (hunk 3 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/PeerMessageBusActor.scala]: At least for me, it doesn't compile otherwise
